core/src/main/java/org/elasticsearch/cockroach/CockroachActionExecutor.java

```diff
@@ -0,0 +1,363 @@
+                                   final TaskManager taskManager, final CockroachTransportService cockroachTransportService,
+                                   final ThreadPool threadPool) {
+        super(settings);
+        this.cockroachActionRegistry = cockroachActionRegistry;
+        this.taskManager = taskManager;
+        this.cockroachTransportService = cockroachTransportService;
+        this.threadPool = threadPool;
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        CockroachTasksInProgress tasks = event.state().custom(CockroachTasksInProgress.TYPE);
+        CockroachTasksInProgress previousTasks = event.previousState().custom(CockroachTasksInProgress.TYPE);
+
+        if (Objects.equals(tasks, previousTasks) == false || event.nodesChanged()) {
+            // We have some changes let's check if they are related to our node
+            String localNodeId = event.state().getNodes().getLocalNodeId();
+            Set<String> notVisitedTasks = new HashSet<>(runningTasks.keySet());
+            if (tasks != null) {
+                for (CockroachTaskInProgress taskInProgress : tasks.entries()) {
+                    if (localNodeId.equals(taskInProgress.getExecutorNode())) {
+                        RunningCockroachTask cockroachTask = runningTasks.get(taskInProgress.getUuid());
+                        if (cockroachTask == null) {
+                            // New task - let's start it
+                            startTask(taskInProgress);
+                        } else {
+                            // The task is still running
+                            notVisitedTasks.remove(taskInProgress.getUuid());
+                            if (taskInProgress.isFinished() == false && cockroachTask.getState() == State.FAILED_NOTIFICATION) {
+                                // We tried to notify the master about this task before but the notification failed and
+                                // the master doesn't seem to know about it - retry notification
+                                restartNotification(cockroachTask);
+                            }
+                        }
+                    }
+                }
+            }
+
+            for (String s : notVisitedTasks) {
+                RunningCockroachTask task = runningTasks.get(s);
+                if (task.getState() == State.NOTIFIED || task.getState() == State.DONE) {
+                    // Result was sent to the caller and the caller acknowledged acceptance of the result
+                    finishTask(s);
+                } else if (task.getState() == State.FAILED_NOTIFICATION) {
+                    // We tried to send result to master, but it failed and master doesn't know about this task
+                    logger.warn("failed to notify master about task {}", task.getUuid());
+                    finishTask(s);
+                }
+                // TODO: task is running locally, but master doesn't know about it
```

Comment:
I'm not following this comment and how the task could be running locally if master didn't have it in the cluster state?